### PR TITLE
RFC: DQMStore index logic rewrite

### PR DESCRIPTION
**Pull request by schneiml**
This PR attempts to complete the complete the conversion from #22157, via #22218 and #22243 et. al. It eliminates `streamId` and `enableMultiThread` from the DQMStore.

This needs to be merged/cherry-picked with #22319 and #22364, it adds another change on top: re-doing the logic that queries MEs out of the set in the DQMStore.

It needs to be discussed if the handling of the runNumber is actually what we want.

This is likely to break multi-run harvesting and online (needs testing).

### CODE REVIEW

File: DQMServices/Components/plugins/MEtoEDMConverter.cc [hunk 1]
```diff
@@ -81,6 +81,8 @@ MEtoEDMConverter::MEtoEDMConverter(const edm::ParameterSet & iPSet) :
   produces<MEtoEDM<TString>, edm::Transition::EndLuminosityBlock>(sName);
 
   consumesMany<DQMToken>();
+  consumesMany<DQMRunToken>();
```
Comment (Dr15Jones) [inline DQMServices/Components/plugins/MEtoEDMConverter.cc]: These need to be 
```
consumesMany<DQMToken, edm::InLumi>();
consumesMany<DQMRunToken, edm::InRun>();
```

File: DataFormats/Histograms/interface/DQMToken.h [hunk 2]
```diff
@@ -25,5 +25,12 @@ class DQMToken
       DQMToken() {}
 };
 
+class DQMRunToken
```
Comment (Dr15Jones) [inline DataFormats/Histograms/interface/DQMToken.h]: There really is not a need for a seperate type just for Run.

File: DQMServices/Components/plugins/MEtoEDMConverter.cc (hunk 1 above)
Comment (fwyzard) [reply DQMServices/Components/plugins/MEtoEDMConverter.cc]: in fact, I guess we could just use `DQMToken` for both Run and Lumi MEs, and let the framework differentiate based on the `InLumi` and `InRun` flags.

File: DataFormats/Histograms/interface/DQMToken.h (hunk 2 above)
Comment (schneiml) [reply DataFormats/Histograms/interface/DQMToken.h]: @Dr15Jones I thought so. But having two `produce` calls with the same Tokens gave me an error, and this was the quick fix.

File: DQMServices/Core/src/DQMEDAnalyzer.cc [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    [this, &run, &setup](DQMStore::IBooker & booker)
+    {
+      booker.cd();
+      this->bookHistograms(booker, run, setup);
+    },
+    run.run(),
+    run.moduleCallingContext()->moduleDescription()->id());
+}
+
+void DQMEDAnalyzer::endRun(edm::Run const& run, edm::EventSetup const& setup)
```
Comment (fwyzard) [inline DQMServices/Core/src/DQMEDAnalyzer.cc]: I would suggest to keep most methods - especially the empty ones - in the .h file.

**Review by fwyzard — CHANGES_REQUESTED** [review]

File: DQMServices/Core/src/DQMEDAnalyzer.cc [hunk 4]
```diff
@@ -0,0 +1,53 @@
+#include "DQMServices/Core/interface/DQMEDAnalyzer.h"
+#include "DataFormats/Histograms/interface/DQMToken.h"
+
+DQMEDAnalyzer::DQMEDAnalyzer() {
+  produces<DQMToken,edm::Transition::EndLuminosityBlock>();
+  produces<DQMRunToken,edm::Transition::EndRun>();
```
Comment (schneiml) [inline DQMServices/Core/src/DQMEDAnalyzer.cc]: @Dr15Jones when I remove the `Run` in this line, I got an framework error, saying I cannot have more than on produce call for the same product. I guess I have to specify both transitions in one call?

File: DQMServices/Components/plugins/MEtoEDMConverter.cc [hunk 5]
```diff
@@ -92,7 +94,7 @@ void
 MEtoEDMConverter::beginJob()
 {
   // Determine if we are running multithreading asking to the DQMStore. Not to be moved in the ctor
-  enableMultiThread_ = dbe->enableMultiThread_;
+  //enableMultiThread_ = dbe->enableMultiThread_;
```
Comment (fwyzard) [inline DQMServices/Components/plugins/MEtoEDMConverter.cc]: just delete the commented-out lines (and drop the empty methods)

File: DQMServices/Components/plugins/MEtoEDMConverter.cc (hunk 1 above)
Comment (schneiml) [reply DQMServices/Components/plugins/MEtoEDMConverter.cc]: ok, let me try. _Maybe_ this fixes our problmes...

File: DQMServices/Components/plugins/MEtoEDMConverter.h [hunk 6]
```diff
@@ -51,8 +51,10 @@
 #include "TObjString.h"
 
 class MEtoEDMConverter : public edm::one::EDProducer<edm::one::WatchRuns,
+                                                     edm::one::WatchLuminosityBlocks,
                                                      edm::EndLuminosityBlockProducer,
-                                                     edm::EndRunProducer>
+                                                     edm::EndRunProducer,
+                                                     edm::one::SharedResources>
```
Comment (fwyzard) [inline DQMServices/Components/plugins/MEtoEDMConverter.h]: what is the purpose of the `SharedResources` ?

**Comment by fwyzard** [comment]
I don't understand one thing: the DQMEDProducer now produces the token, and the METoEDM consumes it.
But what about the DQMEDHarvesters ?

File: DQMServices/Core/src/DQMEDAnalyzer.cc (hunk 4 above)
Comment (Dr15Jones) [reply DQMServices/Core/src/DQMEDAnalyzer.cc]: Hmm. Our framework test works: 
https://github.com/cms-sw/cmssw/blob/ede02fc7b4740a09bb6e578d9ec71129d8c19b19/FWCore/Integration/test/ThingProducer.cc#L14

However, that does use unique product instance names, which was not meant to be required.

File: DQMServices/Components/plugins/MEtoEDMConverter.h (hunk 6 above)
Comment (Dr15Jones) [reply DQMServices/Components/plugins/MEtoEDMConverter.h]: I also do not understand its usage.

File: DQMServices/Core/src/DQMEDAnalyzer.cc (hunk 3 above)
Comment (schneiml) [reply DQMServices/Core/src/DQMEDAnalyzer.cc]: I in general prefer that -- but if it gives me linker errors, I happily put them into a `.cc` file. My theory is that the build system gets more easily confused with implementation in headers (and honestly, I have no idea how this is supposed to work on the linker level, so putting back the `.cc` file was the obvious option).

**Comment by fwyzard** [comment]
Naively I would say that the harvesters should condume only the token(s) of the module(s) that produce the plots they want to harvest; and produce new tokens of their own.

**Comment by Dr15Jones** [comment]
I agree with @fwyzard about the behavior of harvesters

File: DQMServices/Components/plugins/MEtoEDMConverter.cc (hunk 5 above)
Comment (schneiml) [reply DQMServices/Components/plugins/MEtoEDMConverter.cc]: This commit says `HACK`  in the message for a good reason. After #22218, `enableMultiThread` has exactly no reason for existence, but I'll delay removing it everywhere to after #22281 is resolved.

**Comment by fwyzard** [comment]
@dr15jones, how do I make the FastTimerService produce a DQMToken ?
or, should i just call the DQMStore methods at the right moments and let the framework ignore the dependencies ?

File: DQMServices/Components/plugins/MEtoEDMConverter.h (hunk 6 above)
Comment (schneiml) [reply DQMServices/Components/plugins/MEtoEDMConverter.h]: It gives the `usesResource` call, which is a quick way to get a global lock for legacy stuff. Exactly what legacy DQM needs.

Though ideally, the final version of this PR will simply make that all thread save and we can get rid of `usesResource("DQMStore")` everywhere.

**Comment by schneiml** [comment]
@Dr15Jones, @fwyzard -- we agree that we need two types of tokens, one for the "Source" plugins, and one for harvesting. But for now the priority was to get #22281 fixed. This PR is not really done yet, but I wanted to publish the ton of fixes towards fixing ALCA, since likely we'll need others help.

**Comment by Dr15Jones** [comment]
>@Dr15Jones, how do I make the FastTimerService produce a DQMToken ?
or, should i just call the DQMStore methods at the right moments and let the framework ignore the dependencies ?

Services are not allowed to `produce` data. Therefore either
-  talk to the DQMStore directly or
-  have another EDProducer which can get called at the appropriate time?

**Comment by schneiml** [comment]
> Naively I would say that the harvesters should condume only the token(s) of the module(s) that produce the plots they want to harvest; and produce new tokens of their own.

Ah indeed, I mis-read you there. This is where we want to eventually end up, but this will not be possible without subsystem help. Which means the timescale is ~years. 

That would also allow us to specify proper dependencies between harvesters, which currently... they might be there, and scheduled manually, but maybe broken after the harvesting producer migration... We don't really know. (I think there are close to none of these dependencies atm, since it is much more obvious to just do all harvesting in one plugin.)

File: DQMServices/Components/plugins/MEtoEDMConverter.cc (hunk 1 above)
Comment (schneiml) [reply DQMServices/Components/plugins/MEtoEDMConverter.cc]: obviously that would be too easy. Fixed, but #22281 continues...

**Comment by Dr15Jones** [comment]
please test

**Comment by schneiml** [comment]
Btw @Dr15Jones : We also need to get rid of the underscores in the plugin labels of the soon-to-be EDProducers? That would be another big PR (~60 files in subsystem code).

But it seems to mostly work for now, even with underscores...

File: DQMServices/Core/src/DQMEDAnalyzer.cc (hunk 3 above)
Comment (fwyzard) [reply DQMServices/Core/src/DQMEDAnalyzer.cc]: No, the build system does not get "confused".

If you put the implementation of a method inside the class body, e.g.
```c++
class Class {
public:
    void some_method() { ... }
};
```
it is automatically `inline`, and should not create linker problems.

If you put non-inline out-of-body implementation, like
```c++
class Class {
public:
    void some_method();
};

void Class::some_method() { ... }
```
in the .h file, then you do get one version in each .cc file that includes it, and the linker will rightfully compile.

One other thing to pay attention to is if you have virtual methods,  you may need to implement at least one of them in a .cc file, so the compiler knows "where" to emit the vtable and the type information.

I would be curious what problem you ran into here, and if we can find a less drastic solution than moving all methods' implementations to the .cc file.

**Comment by fwyzard** [comment]
> Btw @Dr15Jones : We also need to get rid of the underscores in the plugin labels of the soon-to-be EDProducers? That would be another big PR (~60 files in subsystem code).

I would guess yes, as an underscore in the module label will "confuse" the branch name of the product.

It may work fine as long as you use only `consumesMany`, and do not reference the labels directly.
It may still create problems with the process name, though.

File: DQMServices/Core/src/DQMEDAnalyzer.cc (hunk 3 above)
Comment (schneiml) [reply DQMServices/Core/src/DQMEDAnalyzer.cc]: > One other thing to pay attention to is if you have virtual methods, you may need to implement at least one of them in a .cc file, so the compiler knows "where" to emit the vtable and the type information.

This one. So my gut feeling of putting things into a .cc file was justifed.

And yes, this inline story works fine as long as the methods are not virtual. Then it gets interesting.

**Comment by schneiml** [comment]
This is still WIP and probably does not even compile yet.

**Comment by fwyzard** [comment]
@schneiml , if you agree that this is superseded by #22319, could you close it ?

**Comment by schneiml** [comment]
@fwyzard they are rather diverged by now, because I was doing things that are no longer covered by the title. I'll close it, and make a new PR once theses things are done.

**Pull request closed without merging**